### PR TITLE
Fix https://github.com/dotnet/coreclr/issues/5517

### LINKED ISSUE #5517: Marshal.StructureToPtr: System.AccessViolationException: Attempted to read or write protected memory
Unhandled Exception: System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.Runtime.InteropServices.Marshal.StructureToPtr(Object structure, IntPtr ptr, Boolean fDeleteOld)
   at System.Runtime.InteropServices.Marshal.StructureToPtr[T](T structure, IntPtr ptr, Boolean fDeleteOld)
   at Program.VerifyMarshalLongerByValArrayInAStruct()
   at Program.Main(String[] args)

### PR DESCRIPTION
**Pull request by tijoytom**
Fix #5517  and adding some tests.

### CODE REVIEW

**Comment by tijoytom** [comment]
@yizhang82 @RyanByi FYI

**Comment by RyanByi** [comment]
In MarshalStructure.cs the method name VerifyMarshalLongerByValArrayInAStruct doesn't represent what the actual method does.

**Comment by RyanByi** [comment]
In the underflow and overflow tests the test should fail if no exception is thrown.

**Comment by RyanByi** [comment]
Otherwise looks good.

**Comment by RyanByi** [comment]
Actually one more thing. This is a breaking change. Are we worried about compat at all and should this be only enabled for CoreCLR?

**Comment by yizhang82** [comment]
@RyanByi @tijoytom Let's get a compact review for this one. I'm not worried since this would only impact people who are using it incorrectly.

**Comment by jkotas** [comment]
@dotnet-bot test this please

**Comment by tijoytom** [comment]
@yizhang82 @RyanByi @jkotas I am holding off this till we resolve other issues related to this.

**Comment by tijoytom** [comment]
@yizhang82 @RyanByi PTAL

File: src/vm/fieldmarshaler.cpp [hunk 1]
```diff
@@ -3364,14 +3364,22 @@ VOID FieldMarshaler_FixedStringAnsi::UpdateNativeImpl(OBJECTREF* pCLRValue, LPVO
             nc = m_numchar - 1;
```
Comment (yizhang82) [inline src/vm/fieldmarshaler.cpp]: I think this check is fundamentally incorrect - it did not consider multi-byte ANSI char at all and I believe we should simply let InternalWideToAnsi do the right thing.

File: src/vm/fieldmarshaler.cpp [hunk 2]
```diff
@@ -3364,14 +3364,22 @@ VOID FieldMarshaler_FixedStringAnsi::UpdateNativeImpl(OBJECTREF* pCLRValue, LPVO
...
+            m_BestFitMap,
+            m_ThrowOnUnmappableChar);
+
+        // Handle the case where SizeConst == Number of bytes.For single byte chars 
+        // this will never be the case since nc >= m_numchar check will truncate the last 
+        // character, but for multibyte chars nc>= m_numchar check won't truncate since GetStringLength
+        // gives number of characters but not the actual number of bytes. For such cases need to make 
+        // sure that we dont write one past the buffer.
+        if (cbwritten == m_numchar)
+            --cbwritten;
```
Comment (yizhang82) [inline src/vm/fieldmarshaler.cpp]: This means we'll potentially end up partial characters (for example, half of Chinese) if you overwrite the 2nd half with 0. Why not simply pass m_numchar - 1?

File: src/vm/ilmarshalers.cpp [hunk 3]
```diff
@@ -4385,7 +4384,8 @@ FCIMPL3(void, MngdNativeArrayMarshaler::ConvertContentsToNative, MngdNativeArray
         }
         else
         {
-            pMarshaler->ComToOleArray(pArrayRef, *pNativeHome, pThis->m_pElementMT, pThis->m_BestFitMap, pThis->m_ThrowOnUnmappableChar, pThis->m_NativeDataValid);
+            pMarshaler->ComToOleArray(pArrayRef, *pNativeHome, pThis->m_pElementMT, pThis->m_BestFitMap, 
+                                      pThis->m_ThrowOnUnmappableChar, pThis->m_NativeDataValid, cElements);
```
Comment (yizhang82) [inline src/vm/ilmarshalers.cpp]: nitpick: this is inconsistent in terms of multiple line break with 3599. We should be consistent in terms of coding styles

File: src/vm/ilmarshalers.cpp [hunk 4]
```diff
@@ -4372,10 +4372,9 @@ FCIMPL3(void, MngdNativeArrayMarshaler::ConvertContentsToNative, MngdNativeArray
     if (*pArrayRef != NULL)
     {
         const OleVariant::Marshaler* pMarshaler = OleVariant::GetMarshalerForVarType(pThis->m_vt, TRUE);
-    
+        SIZE_T cElements = (*pArrayRef)->GetNumComponents();
         if (pMarshaler == NULL || pMarshaler->ComToOleArray == NULL)
         {
-            SIZE_T cElements = (*pArrayRef)->GetNumComponents();
             SIZE_T cbArray = cElements;
```
Comment (yizhang82) [inline src/vm/ilmarshalers.cpp]: No need to have the cbArray variable anymore.

File: src/vm/olevariant.cpp [hunk 5]
```diff
@@ -1275,7 +1275,8 @@ void OleVariant::MarshalBoolArrayOleToCom(void *oleArray, BASEARRAYREF *pComArra
 
 void OleVariant::MarshalBoolArrayComToOle(BASEARRAYREF *pComArray, void *oleArray,
                                           MethodTable *pInterfaceMT, BOOL fBestFitMapping,
-                                          BOOL fThrowOnUnmappableChar, BOOL fOleArrayIsValid)
+                                          BOOL fThrowOnUnmappableChar,
+                                          BOOL fOleArrayIsValid,SIZE_T szConst)
```
Comment (yizhang82) [inline src/vm/olevariant.cpp]: nitpick: , SIZE_T

Comment (yizhang82) [inline src/vm/olevariant.cpp]: We should call it cElements. There are other ways to control the size than szConst

File: src/vm/olevariant.cpp [hunk 6]
```diff
@@ -1288,10 +1289,9 @@ void OleVariant::MarshalBoolArrayComToOle(BASEARRAYREF *pComArray, void *oleArra
     CONTRACTL_END;
     
     ASSERT_PROTECTED(pComArray);
-    SIZE_T elementCount = (*pComArray)->GetNumComponents();
 
     VARIANT_BOOL *pOle = (VARIANT_BOOL *) oleArray;
-    VARIANT_BOOL *pOleEnd = pOle + elementCount;
+    VARIANT_BOOL *pOleEnd = pOle + szConst;
```
Comment (yizhang82) [inline src/vm/olevariant.cpp]: Perhaps an contract assert that cElements/szConst <= array->GetNumComponents?

File: src/vm/olevariant.cpp [hunk 7]
```diff
@@ -1365,7 +1365,8 @@ void OleVariant::MarshalWinBoolArrayOleToCom(void *oleArray, BASEARRAYREF *pComA
 
 void OleVariant::MarshalWinBoolArrayComToOle(BASEARRAYREF *pComArray, void *oleArray,
                                           MethodTable *pInterfaceMT, BOOL fBestFitMapping,
-                                          BOOL fThrowOnUnmappableChar, BOOL fOleArrayIsValid)
+                                          BOOL fThrowOnUnmappableChar, 
+                                          BOOL fOleArrayIsValid,SIZE_T szConst)
```
Comment (yizhang82) [inline src/vm/olevariant.cpp]: ditto

File: src/vm/olevariant.cpp [hunk 8]
```diff
@@ -1378,10 +1379,9 @@ void OleVariant::MarshalWinBoolArrayComToOle(BASEARRAYREF *pComArray, void *oleA
     CONTRACTL_END;
     
     ASSERT_PROTECTED(pComArray);
-    SIZE_T elementCount = (*pComArray)->GetNumComponents();
 
     BOOL *pOle = (BOOL *) oleArray;
-    BOOL *pOleEnd = pOle + elementCount;
+    BOOL *pOleEnd = pOle + szConst;
```
Comment (yizhang82) [inline src/vm/olevariant.cpp]: ditto

File: tests/src/Interop/MarshalAPI/MarshalStructure/MarshalStructure.cs [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        var structure1 = new StructWithBoolArray()
+        {
+            array = new bool[]
+            {
+                true,true,true,true
+            }
+        };
+
+        int size = Marshal.SizeOf(structure1);
+        IntPtr memory = Marshal.AllocHGlobal(size);
```
Comment (yizhang82) [inline tests/src/Interop/MarshalAPI/MarshalStructure/MarshalStructure.cs]: I have an idea. Why not allocate additional guard bytes here? We can write a magic number (0xf00df00d, for example) at the end, and make sure that magic number is not overwriten. It is much better than hoping for an AV.

File: tests/src/Interop/SizeConst/SizeConstTest.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+    //UnmanagedType.ByValTStr
+    [DllImport("SizeConstNative",CharSet = CharSet.Ansi)]
+    static extern bool TakeByValTStr(S_CHARArray_ByValTStr s, int size);
+
+    static bool SizeConstByValTStr()
+    {
+        // always marshal managedArray.Length
+        S_CHARArray_ByValTStr s = new S_CHARArray_ByValTStr();
+        s.arr = "有个可爱";
```
Comment (yizhang82) [inline tests/src/Interop/SizeConst/SizeConstTest.cs]: I suspect this doesn't do what you want - our test machines are typically under english code page, so the conversion would simply see them as 8 weird english characters.

File: src/vm/fieldmarshaler.cpp (hunk 2 above)
Comment (tijoytom) [inline src/vm/fieldmarshaler.cpp]: Either way, this is breaking change for the border case. 
  No , this's not breaking change for single byte characters.
Why not simply pass m_numchar - 1
  That won't work for single byte char , for example SizeConst=4 and i am marshaling "ab" , there's no need to truncate 'b'  . 

I have considered all alternatives and came to the conclusion that the current implementation is the one with minimum compact impact as it's limited to  just multi-byte character case when SizeConst ==  Number of bytes.

File: src/vm/fieldmarshaler.cpp (hunk 1 above)
Comment (tijoytom) [inline src/vm/fieldmarshaler.cpp]: Yes , we need this for compact and correctness .

File: tests/src/Interop/SizeConst/SizeConstTest.cs (hunk 10 above)
Comment (tijoytom) [inline tests/src/Interop/SizeConst/SizeConstTest.cs]: Yea , I know that .. I don't see a way to test this without changing machine locale. I guess i can use Windows API SetLocaleInfo info to change the default.

File: src/vm/olevariant.cpp (hunk 5 above)
Comment (tijoytom) [inline src/vm/olevariant.cpp]: Done

File: src/vm/olevariant.cpp (hunk 6 above)
Comment (tijoytom) [inline src/vm/olevariant.cpp]: @yizhang82  Sorry , What's that we want to asset here ?

File: tests/src/Interop/MarshalAPI/MarshalStructure/MarshalStructure.cs (hunk 9 above)
Comment (tijoytom) [inline tests/src/Interop/MarshalAPI/MarshalStructure/MarshalStructure.cs]: Good idea. changed it .

**Comment by tijoytom** [comment]
@yizhang82 @jkotas Updated with review comments , will rebase once CI is clean

**Pull request closed without merging**